hail/src/main/scala/is/hail/lir/Emit.scala

```diff
@@ -146,99 +102,132 @@ object Emit {
+          x.eti.newArray().accept(mv)
         case x: IincX =>
-          mn.instructions.add(new IincInsnNode(getLocalIndex(x.l), x.i))
+          mv.visitIincInsn(getLocalIndex(x.l), x.i)
         case x: StmtOpX =>
-          mn.instructions.add(new InsnNode(x.op))
+          mv.visitInsn(x.op)
         case x: ThrowX =>
-          mn.instructions.add(new InsnNode(ATHROW))
+          mv.visitInsn(ATHROW)
       }
     }
 
     def emitBlock(L: Block): Unit = {
-      mn.instructions.add(labelNodes(L))
+      mv.visitLabel(labels(L))
       var x = L.first
       while (x != null) {
         emitX(x, 0)
         x = x.next
       }
     }
 
-    mn.instructions.add(start)
+    mv.visitLabel(start)
 
     emitBlock(m.entry)
     for (b <- blocks) {
       if (b ne m.entry)
         emitBlock(b)
     }
 
-    mn.instructions.add(end)
+    mv.visitLabel(end)
 
-    mn.maxStack = maxStack
-  }
+    for (l <- locals) {
+      if (!l.isInstanceOf[Parameter]) {
+        val n = localIndex(l)
+        val name = if (l.name == null) s"local$n" else l.name
+        mv.visitLocalVariable(name, l.ti.desc, null, start, end, n)
+      }
+    }
 
-  def apply(c: Classx[_], print: Option[PrintWriter]): Array[Byte] = {
-    val cn = new ClassNode()
+    mv.visitMaxs(maxStack, nLocals)
 
-    cn.version = V1_8
-    cn.access = ACC_PUBLIC
+    mv.visitEnd()
 
-    c.sourceFile.foreach(cn.sourceFile = _)
+    instructionCount
+  }
 
-    cn.name = c.name
-    cn.superName = c.superName
-    for (intf <- c.interfaces)
-      cn.interfaces.add(intf)
+  def emitClass(c: Classx[_], cv: ClassVisitor, logMethodSizes: Boolean): Unit = {
+    cv.visit(V1_8, ACC_PUBLIC, c.name, null, c.superName, c.interfaces.toArray)
+    c.sourceFile.foreach(cv.visitSource(_, null))
 
     for (f <- c.fields) {
-      val fn = f match {
-        case f: Field => new FieldNode(ACC_PUBLIC, f.name, f.ti.desc, null, null)
-        case f: StaticField => new FieldNode(ACC_PUBLIC | ACC_STATIC, f.name, f.ti.desc, null, null)
+      f match {
+        case f: Field => cv.visitField(ACC_PUBLIC, f.name, f.ti.desc, null, null)
+        case f: StaticField => cv.visitField(ACC_PUBLIC | ACC_STATIC, f.name, f.ti.desc, null, null)
       }
-      cn.fields.add(fn)
     }
 
     for (m <- c.methods) {
-      emit(cn, m)
+      val instructionCount = emitMethod(cv, m, c.sourceFile.isDefined)
+      if (logMethodSizes) {
+        log.info(s"instruction count: $instructionCount: ${ c.name }.${ m.name }")
+        if (instructionCount > 8000)
+          log.warn(s"big method: $instructionCount: ${ c.name }.${ m.name }")
+      }
     }
 
-    asBytes(cn, print)
+    cv.visitEnd()
+  }
+
+  def apply(c: Classx[_], print: Option[PrintWriter]): Array[Byte] = {
+    val bytes = try {
+      val cw = new ClassWriter(ClassWriter.COMPUTE_MAXS + ClassWriter.COMPUTE_FRAMES)
+
+      emitClass(c, cw, logMethodSizes = true)
+
+      val b = cw.toByteArray
+      // This next line should always be commented out!
```

Comment:
Is it for debugging? Maybe just say why I would want to uncomment it?

Reply:
This has been here a long time; I just moved it from `asBytes`. But I can add a bit of explanation.